From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Gabriel Roman <gabrielroman@google.com>
Date: Thu, 4 Jun 2026 02:01:27 -0700
Subject: [PATCH 2/9] Truncate long device admin descriptions

Limit the length of the description loaded from a device admin's
manifest to a maximum of 8192 characters. This prevents potential issues
with excessively long descriptions.

Bug: 492093013
Test: manual
Flag: EXEMPT BUGFIX
Cherrypick-From: https://googleplex-android-review.googlesource.com/q/commit:f1ed046d2c48a38b9a800ef6a86cc14d3ab312a3
Security vulnerability cherry pick for https://b.corp.google.com/issues/492093013
Topic: https://googleplex-android-review.git.corp.google.com/q/topic:%22b492093013%22
Cherry picked the main CL for reordering
#b/492093013
Cherrypick-From: https://googleplex-android-review.googlesource.com/q/commit:8f743f0c5842ca36f1b6fdce6a3646a65e84caab
Merged-In: I06b621e37d97fd8554172c2446e2838dfaccc104
Change-Id: I06b621e37d97fd8554172c2446e2838dfaccc104

(cherry picked from commit 96d0c52e13bf0f58e1f0b1fc3687f137d0ae1122)
---
 .../android/app/admin/DeviceAdminInfo.java    | 11 ++++++
 .../android/content/pm/ComponentInfo.java     | 39 +++++++++++++++----
 2 files changed, 42 insertions(+), 8 deletions(-)

diff --git a/core/java/android/app/admin/DeviceAdminInfo.java b/core/java/android/app/admin/DeviceAdminInfo.java
index 986205a346f7..1a909e309dde 100644
--- a/core/java/android/app/admin/DeviceAdminInfo.java
+++ b/core/java/android/app/admin/DeviceAdminInfo.java
@@ -465,6 +465,17 @@ public final class DeviceAdminInfo implements Parcelable {
         throw new NotFoundException();
     }
 
+    /**
+     * @hide
+     * Load user-visible description associated with this device admin. The description is trimmed
+     * to 1024 characters by leveraging {@link ComponentInfo#loadDescription(PackageManager)}.
+     *
+     * @param pm Supply a PackageManager used to load the device admin's resources.
+     */
+    public CharSequence loadDescriptionSafe(@NonNull PackageManager pm) throws NotFoundException {
+        return mActivityInfo.loadDescription(pm);
+    }
+
     /**
      * Load the user-displayed icon for this device admin.
      *
diff --git a/core/java/android/content/pm/ComponentInfo.java b/core/java/android/content/pm/ComponentInfo.java
index ff48ffafba8a..e9ffcc6da30a 100644
--- a/core/java/android/content/pm/ComponentInfo.java
+++ b/core/java/android/content/pm/ComponentInfo.java
@@ -16,6 +16,7 @@
 
 package android.content.pm;
 
+import android.annotation.NonNull;
 import android.annotation.SuppressLint;
 import android.app.Activity;
 import android.app.Service;
@@ -24,11 +25,15 @@ import android.content.BroadcastReceiver;
 import android.content.ComponentName;
 import android.content.ContentResolver;
 import android.content.Context;
+import android.content.res.Resources.NotFoundException;
 import android.graphics.drawable.Drawable;
 import android.os.Build;
 import android.os.Parcel;
+import android.text.TextUtils;
 import android.util.Printer;
 
+import java.util.Objects;
+
 /**
  * Base class containing information common to all application components
  * ({@link ActivityInfo}, {@link ServiceInfo}).  This class is not intended
@@ -44,7 +49,7 @@ public class ComponentInfo extends PackageItemInfo {
      * part of.
      */
     public ApplicationInfo applicationInfo;
-    
+
     /**
      * The name of the process this component should run in.
      * From the "android:process" attribute or, if not set, the same
@@ -83,7 +88,7 @@ public class ComponentInfo extends PackageItemInfo {
      * attribute or, if not set, 0.
      */
     public int descriptionRes;
-    
+
     /**
      * Indicates whether or not this component may be instantiated.  Note that this value can be
      * overridden by the one in its parent {@link ApplicationInfo}.
@@ -105,6 +110,9 @@ public class ComponentInfo extends PackageItemInfo {
      */
     public boolean directBootAware = false;
 
+    /** @hide */
+    public static final int MAX_SAFE_DESCRIPTION_LENGTH = 1024;
+
     public ComponentInfo() {
     }
 
@@ -145,18 +153,33 @@ public class ComponentInfo extends PackageItemInfo {
         return name;
     }
 
+    /** @hide */
+    public CharSequence loadDescription(@NonNull PackageManager pm) {
+        Objects.requireNonNull(pm);
+        if (descriptionRes != 0) {
+            try {
+                return TextUtils.trimToSize(
+                        pm.getText(packageName, descriptionRes, applicationInfo),
+                        MAX_SAFE_DESCRIPTION_LENGTH);
+            } catch (OutOfMemoryError e) {
+                throw new NotFoundException();
+            }
+        }
+        throw new NotFoundException();
+    }
+
     /**
      * Return whether this component and its enclosing application are enabled.
      */
     public boolean isEnabled() {
         return enabled && applicationInfo.enabled;
     }
-    
+
     /**
      * Return the icon resource identifier to use for this component.  If
      * the component defines an icon, that is used; else, the application
      * icon is used.
-     * 
+     *
      * @return The icon associated with this component.
      */
     public final int getIconResource() {
@@ -173,7 +196,7 @@ public class ComponentInfo extends PackageItemInfo {
     public final int getLogoResource() {
         return logo != 0 ? logo : applicationInfo.logo;
     }
-    
+
     /**
      * Return the banner resource identifier to use for this component. If the
      * component defines a banner, that is used; else, the application banner is
@@ -242,7 +265,7 @@ public class ComponentInfo extends PackageItemInfo {
         dest.writeInt(exported ? 1 : 0);
         dest.writeInt(directBootAware ? 1 : 0);
     }
-    
+
     protected ComponentInfo(Parcel source) {
         super(source);
         applicationInfo = ApplicationInfo.CREATOR.createFromParcel(source);
@@ -263,7 +286,7 @@ public class ComponentInfo extends PackageItemInfo {
     public Drawable loadDefaultIcon(PackageManager pm) {
         return applicationInfo.loadIcon(pm);
     }
-    
+
     /**
      * @hide
      */
@@ -280,7 +303,7 @@ public class ComponentInfo extends PackageItemInfo {
     protected Drawable loadDefaultLogo(PackageManager pm) {
         return applicationInfo.loadLogo(pm);
     }
-    
+
     /**
      * @hide
      */
